### PR TITLE
[Console] Add mbstring extension support for OutputFormatter

### PR DESCRIPTION
**Pull request by dsxack**
| Q | A |
| --- | --- |
| Bug fix? | [no] |
| New feature? | [yes] |
| BC breaks? | [no] |
| Deprecations? | [no] |
| Fixed tickets | [#12530] |
| License | MIT |
| Doc PR | - |

Fix for this issue:
#12530

Related pull-request #14348

### CODE REVIEW

**Comment by Tobion** [comment]
We don't support mb string function overloading. It's in the symfony requirements. Otherwise we have to adapt much more places, like routing etc.

File: src/Symfony/Component/Console/Formatter/OutputFormatter.php [hunk 1]
```diff
@@ -238,4 +238,22 @@ private function applyCurrentStyle($text)
...
+    /**
+     * Get part of string with mbstring extension support.
+     *
+     * @param string $str    The string being checked.
+     * @param int    $start  The first position used in str.
+     * @param int    $length [optional] The maximum length of the returned string.
+     *
+     * @return string String the extracted part of string or false on failure.
+     */
+    private function substr($str, $start, $length = null)
```
Comment (phansys) [inline src/Symfony/Component/Console/Formatter/OutputFormatter.php]: This method could be static since object instance isn't used at all.

**Comment by DsXack** [comment]
@Tobion, this hack is one of the requirements of the most popular commercial CMS, which is used in Russia - Bitrix. Refusal to support a mbstring simfony-components means a complete incompatibility of the CMS with all the tools built on symfony-components

**Comment by Tobion** [comment]
`mbstring.func_overload` is a __requirement__ of this CMS? They should fix their code. What value does it require?

**Comment by javiereguiluz** [comment]
@Tobion what you said about function overload is correct, but could you please reconsider this proposal?

Keep in mind that the focus of this PR is very limited (just the Console component) and the introduced changes are minimal but produce a huge impact (from not being able to use the component to happily use it).

Lastly, please consider that the Symfony Console component is very special. It's wildly used outside Symfony (it has more than 8.5 million downloads on Packagist!)

**Comment by Tobion** [comment]
@javiereguiluz Then we should make it compatible everywhere. And remove the symfony requirement.

**Comment by DsXack** [comment]
CMS is a system of commercial closed-source. Their developers do not explain the reasons for their decisions. But for this CMS largest Russian hosting-providers have special tariffs that include specific web server configuration, including the mbstring. Sometimes it is very hard to change this setting, in some situations have to move the project client to another server

**Comment by javiereguiluz** [comment]
@Tobion could you make an estimate of the impact of this change in the entire Symony codebase or it's impossible to know? Should we change a lot of things or it's a matter of minor tweaks here and there?

**Comment by Tobion** [comment]
I think it should only be where we use regex matching together with any `str...` operation. Or were we access a string by byte offset `$str[$pos]` which would also be the wrong index.

**Comment by DsXack** [comment]
Overloaded functions list: http://php.net/manual/en/mbstring.overload.php

required value of mbstring.func_overload is 2 for CMS Bitrix

**Comment by Tobion** [comment]
So most uses of `substr` etc should be safe as they would behave the same. But as soon as you combine it with parsing, e.g. `preg_match_all`, the assumpion that we work with bytes doesn't hold anymore.

**Comment by DsXack** [comment]
The decision to ban mbstring was taken in #6043.

Well, then the only solution that I can see to somehow solve this problem - to attempt to influence the Bitrix. And in the future - RFC request for PHP for deprecate mbstring.func_overload. Thank you for your time.

**Comment by petrgrishin** [comment]
I agree that this is a problem Bitrix.
@DsXack I approve of your decision.

**Comment by DsXack** [comment]
@Tobion, @javiereguiluz, Please, help to pay attention to the developers php. No one answered, but the problem exists.

https://bugs.php.net/bug.php?id=65785

**Comment by Tobion** [comment]
Please create a new PR for reconsideration as this cannot be reopended as the branch has been modified.

**Comment by Tobion** [comment]
And please provide a performance comparison to ensure it does not decrease too much with these changes.

**Comment by DsXack** [comment]
I think it should be changed not simfony. I think mbstring.func_overload should be deprecated in mbstring extension, in PHP. But I do not know where to write, to pay attention to this php developers.

**Comment by DsXack** [comment]
please vote issue "mbstring.func_overload should be depreciated." https://bugs.php.net/bug.php?id=65785

**Pull request closed without merging**